equational_theories/Completeness.lean

```diff
@@ -0,0 +1,219 @@
+-- Statement and proof of the completeness theorem for magmas,
+-- we roughly follow Baader & Nipkow's "Term Rewriting and All That",
+-- and
+import equational_theories.FreeMagma
+import Mathlib.Data.Set.Defs
+
+#check Set
+
+structure MagmaLaw (α : Type) where
+  lhs : FreeMagma α
+  rhs : FreeMagma α
+
+infix:60 " ≃ " => MagmaLaw.mk
+
+def substFreeMagma {α β} (t : FreeMagma α) (σ : α → FreeMagma β) : FreeMagma β :=
+match t with
+| .Leaf a => σ a
+| t₁ ⋆ t₂ => substFreeMagma t₁ σ ⋆ substFreeMagma t₂ σ
+
+infix:66 " ⬝ " => substFreeMagma
+
+#check ((Lf 0) ⬝ (λ i ↦ Lf i))
+
+@[inline, simp]
+def Ctx α := Set (MagmaLaw α)
+
+#print Membership
+
+#print Set.instMembership
+
+-- FIXME: figure out how to remove this.
+instance Ctx.Membership α : Membership (MagmaLaw α) (Ctx α) := ⟨ Set.instMembership.mem ⟩
+
+section DeriveDef
+
+set_option hygiene false
+
+local infix:50 " ⊢ " =>  derive
+
+-- We keep this in type, because we're going to want to gather
+-- the (finite!) set of required axioms later.
+inductive derive {α} : Ctx α → MagmaLaw α → Type :=
+| Ax Γ E (h : E ∈ Γ) : Γ ⊢ E
+| Ref Γ t : Γ ⊢ (t ≃ t)
+| Sym Γ t u : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ t)
+| Trans Γ t u v : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ v) → Γ ⊢ (t ≃ v)
+-- This is not as polymorphic as it could be, shouldn't be an issue at the moment
+| Subst Γ t u σ : Γ ⊢ (t ≃ u) → Γ ⊢ (t ⬝ σ ≃ u ⬝ σ)
```

Comment:
Why do you need this? Isn't it just applying `Cong` with one of the sides being `Ref`?